### DISCUSSION TITLE
Running a Custom Processor standalone

### DISCUSSION THREAD

**Discussion by macmule** [autopkg/autopkg#952] [General]
Hi folks,

Has anyone recently ran a Custom Processor standalone?

I've been following the steps at - [https://github.com/autopkg/autopkg/wiki/Developing-Custom-Processors#running-a-custom-processor-standalone](https://github.com/autopkg/autopkg/wiki/Developing-Custom-Processors#running-a-custom-processor-standalone), and hit a few issues.

Like, the whole process hanging.. which commenting out [https://github.com/autopkg/autopkg/blob/8a12727820da6be40ee27530ac492c30283e54ca/Code/autopkglib/__init__.py#L651 ](https://github.com/autopkg/autopkg/blob/8a12727820da6be40ee27530ac492c30283e54ca/Code/autopkglib/__init__.py#L651 )resolves.. the leading to a traceback..
```
Traceback (most recent call last):
  File "/Users/ben.toms/Library/AutoPkg/RecipeRepos/com.github.autopkg.nstrauss-recipes/VirusTotalReporter/VirusTotalReporter.py", line 376, in <module>
    PROCESSOR.execute_shell()
  File "/Library/AutoPkg/autopkglib/__init__.py", line 652, in execute_shell
    self.parse_arguments()
  File "/Library/AutoPkg/autopkglib/__init__.py", line 603, in parse_arguments
    update_data(self.env, key, value)
  File "/Library/AutoPkg/autopkglib/__init__.py", line 453, in update_data
    a_dict[key] = do_variable_substitution(value)
TypeError: 'NoneType' object does not support item assignment
```

Which, substantiating the dict resolves..

So, before I go further into this rabbit hole I wanted to be see if it's me holding things wrong or if indeed there are issues to be sorted.

**Comment by grahampugh** [comment]
I run custom processors alone and it works (see `jamf-upload.sh`). Looks to me like you're missing a required variable assignment.

**Reply by macmule** [reply]
I'm about to write a very simple test processor.. as the two I've tried hit the same issues..
But, what's the simplest and quickest way for me to run one of yours?

https://github.com/grahampugh/jamf-upload/blob/main/jamf-upload.sh#L1137

it might even just be an issue with how I'm calling python.. as seems one difference (except plist vs args) in you jamf-upload.sh... from what I'm doing:

```
export PYTHONPATH=/Library/AutoPkg; /usr/local/autopkg/python /Users/ben.toms/Library/AutoPkg/RecipeRepos/com.github.autopkg.nstrauss-recipes/VirusTotalReporter/VirusTotalReporter.py VIRUSTOTAL_ALWAYS_REPORT=True VIRUSTOTAL_SUBMIT_NEW=True
```

**Reply by macmule** [reply]
Sadly no luck..
`export PYTHONPATH=/Library/AutoPkg; /Library/AutoPkg/Python3/Python.framework/Versions/Current/bin/python3 /Users/ben.toms/Library/AutoPkg/RecipeRepos/com.github.autopkg.nstrauss-recipes/VirusTotalReporter/VirusTotalReporter.py pathname=/Users/ben.toms/Library/AutoPkg/Cache/com.github.autopkg.download.VLC/VLC-3.0.21-Install-universal.dmg VIRUSTOTAL_ALWAYS_REPORT=True VIRUSTOTAL_SUBMIT_NEW=True`

The above just hangs.. I knock up a simple processor

**Reply by macmule** [reply]
Forgot to say thanks @grahampugh, helped lead me to the issue :)

**Comment by macmule** [comment]
So.. seems like this might be an issue when trying to use command line arguments (as per: [https://github.com/autopkg/autopkg/wiki/Developing-Custom-Processors#sample-usage---command-line-argumentsvariables](https://github.com/autopkg/autopkg/wiki/Developing-Custom-Processors#sample-usage---command-line-argumentsvariables))

The below change to the execute_shell function in the __init__.py ([https://github.com/autopkg/autopkg/blob/master/Code/autopkglib/__init__.py#L647]https://github.com/autopkg/autopkg/blob/master/Code/autopkglib/__init__.py#L647), seems to help:
```
    def execute_shell(self):
        """Execute as a standalone binary on the commandline."""

        try:
            if not sys.argv[1:]:
                self.read_input_plist()
            else:
                self.parse_arguments()
```

After which hitting a traceback, which might be processor related:

```
Traceback (most recent call last):
  File "/Users/ben.toms/Library/AutoPkg/RecipeRepos/com.github.autopkg.nstrauss-recipes/VirusTotalReporter/VirusTotalReporter.py", line 377, in <module>
    PROCESSOR.execute_shell()
  File "/Library/AutoPkg/autopkglib/__init__.py", line 656, in execute_shell
    self.parse_arguments()
  File "/Library/AutoPkg/autopkglib/__init__.py", line 603, in parse_arguments
    update_data(self.env, key, value)
  File "/Library/AutoPkg/autopkglib/__init__.py", line 453, in update_data
    a_dict[key] = do_variable_substitution(value)
TypeError: 'NoneType' object does not support item assignment
```

**Comment by macmule** [comment]
So now the fun question.. should I bother trying to get command line usage/arguments working again?

Or, should we just remove the `parse_arguments` def and only support plist input?

**Comment by macmule** [comment]
Raised a PR for this - https://github.com/autopkg/autopkg/pull/953

**Comment by homebysix** [comment]
The PR should be included in the next version, expected to be 2.7.6. Thanks!
